### ISSUE TITLE
Provide an option to bypass reshape on boot

### ISSUE THREAD

**Issue by raphaelsc** [scylladb/scylla#7738]
Let's allow user to make node online ASAP even though that potentially mean performance consequences later on. But it's important to allow user to control this.

--bypass-reshape-on-boot (disabled by default)

/cc @avikivity @slivne @bhalevy

**Comment by avikivity** [comment]
I don't see how it will be useful. People notice reshape only when it starts, by then it is too late.

Also, without reshape, things might not work too well.

**Comment by raphaelsc** [comment]
> I don't see how it will be useful. People notice reshape only when it starts, by then it is too late.
> 
> Also, without reshape, things might not work too well.

With reshape, we opt for optimal read perf, from the moment node is online, at cost of increasing boot time. Without reshape, we potentially have suboptimal read perf until compaction catches up. We made the former the default behavior. However, reshape can take even hours to complete. Users that MUST have the node online ASAP will need the latter option, not a question of preference but of needs. If the users notice reshape is taking too long, they can shutdown scylla and feed the bypass option.

**Comment by dorlaor** [comment]
I agree it can be helpful and sure that long reshape will surprise users. 
You can even change the cmdline option to a restapi control so a restart
wouldn't be needed. Worth to add a log with the option so when a reshape
begins, the user will see the bypass option in the log

**Comment by slivne** [comment]
Disabling resharding / reshaping is more complex then simply allowing boot to continue - there are implications to that - actually allowing reads to use shared / none reshaped sstables.

We have code that assumes at the end of the boot that shared ssatables do not exist and reshaping has ended (293d4117c1950f9dc3cf943de119c1bbe6202fca)

There may be more issues related to that - 

It would be good if we provided more info on boot about how much data needs reshaping / how long it will take - and even better if we can indicate this will happen before the upgrade. 

Reshaping doesn't grantee only optimal read performance it also grantees data will get expired and use will not run out of disk space (TWCS sstables not conforming to window will be held longer).

If shaping code is part of 4.0 and we have major compaction supporting that we could recommend users to run major compaction prior to upgrade for certain tables (e.g. TWCS).

**Comment by raphaelsc** [comment]
> Disabling resharding / reshaping is more complex then simply allowing boot to continue - there are implications to that - actually allowing reads to use shared / none reshaped sstables.
> 

Btw, I am not suggesting to bypass resharding as that would have serious complications. Previously, we launched resharding in background but we can no longer do that. So the suggestion is to only bypass reshape.

> We have code that assumes at the end of the boot that shared ssatables do not exist and reshaping has ended ([293d411](https://github.com/scylladb/scylla/commit/293d4117c1950f9dc3cf943de119c1bbe6202fca))
> 
> There may be more issues related to that -
> 
> It would be good if we provided more info on boot about how much data needs reshaping / how long it will take - and even better if we can indicate this will happen before the upgrade.
> 
> Reshaping doesn't grantee only optimal read performance it also grantees data will get expired and use will not run out of disk space (TWCS sstables not conforming to window will be held longer).

Regular compaction can also be described as an incremental reshape, which works towards reshaping the data in a table to conform to the strategy invariant. It enables the interposer consumer (i.e. segregation or whatever) if it finds that the compacting data isn't conforming to the invariant.

Let's say I am an user that need my node online ASAP, and I am willing to allow compaction to incrementally reshape my data. Then I have an option to doso. 

The default behavior would still be: reshape my data before node is online.
Enabling this user would be: reshape my data incrementally after node is online through regular compaction.

So it's a trade off.

What we're willing to do, with this option, is to allow users to explicitly go with the same behavior as before. The new behavior will be kept the default.


> 
> If shaping code is part of 4.0 and we have major compaction supporting that we could recommend users to run major compaction prior to upgrade for certain tables (e.g. TWCS).

**Comment by slivne** [comment]
@avikivity are you ok with adding a flag to bypass reshaping on boot - if the user selects that ?

**Comment by avikivity** [comment]
I guess. I can't say I'm happy with adding more and more options.

**Comment by bhalevy** [comment]
More options means proliferation of test configurations

**Comment by slivne** [comment]
@bhalevy please queue it after the urgent items RaphaelSC has

**Comment by gnumoreno** [comment]
User reported on slack:
```
I am in the process of upgrading my cluster from 4.1.11 to 4.2.3 and may have hit a blocker. I deployed the upgrade on our test db and there was ~1hr downtime per node as the init process did a reshape to almost all of our data on disk. During this time, the node is completely down. If I were to apply this upgrade to the prod db, at the same throughput I figure the reshape would take ~48 hrs. Is there a way to delay or bypass the new reshape at init feature?
```
`https://scylladb-users.slack.com/archives/C2NLNBXLN/p1614286305066000`

**Comment by bhalevy** [comment]
> The default behavior would still be: reshape my data before node is online.
> Enabling this option the behavior would be: reshape my data incrementally after node is online through regular compaction.

It's unclear why the default behavior should be to reshape on (re)start, sacrificing availability (the node is not serving requests until reshape is done) for potential performance benefits.

Let's look at possible reasons the sstables need reshaping:
1. Compaction went behind before the node restarted and backlog accumulated.
This is a backlog-controller problem and we should fix that instead of "catching up" after restart.
2. Restart during/right after repair / repair-based node operations.
Soon with off-strategy compaction, we can detect the repaired sstables in the distributed_loader using their sstable_origin metadata, put them in the maintenance sstable set, and resume off-strategy compaction.
3. Restart after compaction-strategy change.
We plan to run off-strategy compaction after such event, and it too should be resumed in the background after restart.

The bottom line is that it's unreasonable to delay to node time-to-serve by more than a few minutes, therefore it would be better to disable reshape on boot, and if we introduce the proposed option, for emergency, unforeseen reasons, let's make it enable-reshape-on-boot and have it default to `false`.

**Comment by avikivity** [comment]
> > The default behavior would still be: reshape my data before node is online.
> > Enabling this option the behavior would be: reshape my data incrementally after node is online through regular compaction.
> 
> It's unclear why the default behavior should be to reshape on (re)start, sacrificing availability (the node is not serving requests until reshape is done) for potential performance benefits.

It's because unreshaped data has terrible performance. The node is up, but the request rate and latency are terrible. We've seen the same with resharding.


> 
> Let's look at possible reasons the sstables need reshaping:
> 
> 1. Compaction went behind before the node restarted and backlog accumulated.
>    This is a backlog-controller problem and we should fix that instead of "catching up" after restart.
> 2. Restart during/right after repair / repair-based node operations.
>    Soon with off-strategy compaction, we can detect the repaired sstables in the distributed_loader using their sstable_origin metadata, put them in the maintenance sstable set, and resume off-strategy compaction.
> 3. Restart after compaction-strategy change.
>    We plan to run off-strategy compaction after such event, and it too should be resumed in the background after restart.
> 
> The bottom line is that it's unreasonable to delay to node time-to-serve by more than a few minutes, therefore it would be better to disable reshape on boot, and if we introduce the proposed option, for emergency, unforeseen reasons, let's make it enable-reshape-on-boot and have it default to `false`.

It's also unreasonable to restart after a repair and have a performance drop that takes us below what the user expects.

A repair or "nodetool refresh" that happens online gets reshaped in the maintenance group, and so doesn't affect performance. But after a restart we lose the distinction between these sstables.

**Comment by avikivity** [comment]
I think an option is not a good solution for this. The flow is terrible:

1. start the node, notice a reshape taking too long, decide to stop it
2. edit configuration
3. restart
4. remember to edit configuration again.

It should be a nodetool command:

1. start the node, notice a reshape taking too long, decide to stop it
2. issue `nodetool reshape abort`

**Comment by raphaelsc** [comment]
On Wed, Mar 3, 2021 at 8:19 AM Avi Kivity <notifications@github.com> wrote:

> I think an option is not a good solution for this. The flow is terrible:
>
>    1. start the node, notice a reshape taking too long, decide to stop it
>    2. edit configuration
>    3. restart
>    4. remember to edit configuration again.
>
> It should be a nodetool command:
>
>    1. start the node, notice a reshape taking too long, decide to stop it
>    2. issue nodetool reshape abort
>
> That can be done with the existing nodetool stop, so it becomes nodetool
stop reshape. I need to make it work as nodetool stop is currently only
working for regular and cleanup.

**Comment by STRATZ-Ken** [comment]
I second this, I have been down for about 24 hours, and expect an outage of about 4 days while the data reshapes. :(

**Comment by vladzcloudius** [comment]
I think this feature ("reshaping") should be disabled by default to begin with.
I also think that this feature in its current form is simply put useless. See #8844 for more details.

**Comment by avikivity** [comment]
> I second this, I have been down for about 24 hours, and expect an outage of about 4 days while the data reshapes. :(

Why do you have an outage? You should update nodes one by one.

**Comment by vladzcloudius** [comment]
> > > The default behavior would still be: reshape my data before node is online.
> > > Enabling this option the behavior would be: reshape my data incrementally after node is online through regular compaction.
> > 
> > 
> > It's unclear why the default behavior should be to reshape on (re)start, sacrificing availability (the node is not serving requests until reshape is done) for potential performance benefits.
> 
> It's because unreshaped data has terrible performance. The node is up, but the request rate and latency are terrible. We've seen the same with resharding.
> 
> > Let's look at possible reasons the sstables need reshaping:
> > 
> > 1. Compaction went behind before the node restarted and backlog accumulated.
> >    This is a backlog-controller problem and we should fix that instead of "catching up" after restart.
> > 2. Restart during/right after repair / repair-based node operations.
> >    Soon with off-strategy compaction, we can detect the repaired sstables in the distributed_loader using their sstable_origin metadata, put them in the maintenance sstable set, and resume off-strategy compaction.
> > 3. Restart after compaction-strategy change.
> >    We plan to run off-strategy compaction after such event, and it too should be resumed in the background after restart.
> > 
> > The bottom line is that it's unreasonable to delay to node time-to-serve by more than a few minutes, therefore it would be better to disable reshape on boot, and if we introduce the proposed option, for emergency, unforeseen reasons, let's make it enable-reshape-on-boot and have it default to `false`.
> 
> It's also unreasonable to restart after a repair and have a performance drop that takes us below what the user expects.
> 
> A repair or "nodetool refresh" that happens online gets reshaped in the maintenance group, and so doesn't affect performance. But after a restart we lose the distinction between these sstables.

@avikivity I'm not sure I'm following.
This "reshaping" we are discussing has very little to do with reshaping that happens when you change to LCS for instance.
The "reshaping" in question is simply a process that tries to clear a compaction backlog.

AFAIK we don't distinguish between data that came from repair, `nodetool refresh` or any other source when we compact. Hence the performance is not supposed to be worse after you start the node than it was before it was stopped.

The only exception is if a compaction strategy has been changed when the node was DOWN and the real reshaping needs to take place. And this is the only use case I can imagine when the current behavior can be useful and hence we should allow to explicitly enabling it as suggested by @bhalevy .

In all other cases we are simply looking at a compaction backlog which doesn't justify delaying a boot like @bhalevy explained above: https://github.com/scylladb/scylla/issues/7738#issuecomment-789481406

**Comment by bhalevy** [comment]
@raphaelsc Reading the code, it looks like `nodetool stop RESHAPE` could work if the REST api for it would start early enough, before the distributed loader initializes the keyspaces, right? (compaction_manager is already enabled since we need it to run reshard or reshape when loading the stables)

**Comment by raphaelsc** [comment]
On Mon, Jul 26, 2021 at 11:24 AM Benny Halevy ***@***.***>
wrote:

> @raphaelsc <https://github.com/raphaelsc> Reading the code, it looks like nodetool
> stop RESHAPE could work if the REST api for it would start early enough,
> before the distributed loader initializes the keyspaces, right?
> (compaction_manager is already enabled since we need it to run reshard or
> reshape when loading the stables)
>

we need to make the command stop the whole reshape procedure, not only a
single individual reshape compaction job. I am planning to work on it
today. then we've to patch nodetool to accept the RESHAPE type.

**Comment by bhalevy** [comment]
> then we've to patch nodetool to accept the RESHAPE type.

Are you sure it's validating the compaction type?
I didn't see that in the code.  It seems like it's just passing any string given on the command line, but maybe I missed it.

**Comment by raphaelsc** [comment]
On Mon, Jul 26, 2021 at 11:41 AM Benny Halevy ***@***.***> wrote:
>
> then we've to patch nodetool to accept the RESHAPE type.
>
> Are you sure it's validating the compaction type?
> I didn't see that in the code. It seems like it's just passing any string given on the command line, but maybe I missed it.

I remember it will only accept compaction types that are defined. I'll
double check it, and if adding reshape type is needed, I'll send a
patch for nodetool.

**Comment by raphaelsc** [comment]
reasons to reshape on boot:

1) restarted during repair and before offstrategy completed, so sstables with repair origin need reshape before integration.
2) resharding triggered, so resulting sstables need reshape before integration

NOTE: reshape on boot isn't needed for refresh, because sstables are only moved out of upload dir after reshaped in place first.

what if we limit reshape on boot only for reason 2?

for reason 1, we can make boot add sstables with repair origin to maintenance set and trigger offstrategy compaction (can be thought of as an asynchronous reshape which doesn't block boot), so node can be online much faster.


the other option is to always run reshape on boot if needed and user aborts it if taking too long. unfortunately, not even a subset of API is available at early boot stage, so I'll have to refactor a few things for this. 


@avikivity, @bhalevy  what do you think?

**Comment by bhalevy** [comment]
@raphaelsc I think that loading the sstables with repair origin to the maintenance is a good idea in all scenarios  whether they are reshaped on startup or just arm the 5 minutes timer, to let the manager a chance to resume repair in case the node restarted mid-repair.

To keep things simple, assuming we start listening on the REST api before initializing the user tables, we can do the reshaping during startup also after repair, and allow stopping it via the api.  Then, in this case, we can use the offstategy compaction timer to resume it later.

**Comment by raphaelsc** [comment]
For those who were following this work, reshape can now be aborted on boot.

The flow is as follow:
1) Reshape starts on boot.
2) User notices it's taking too long.
3) nodetool stop RESHAPE.
4) Node aborts reshape, while preserving compaction work completed up to this point, and become online ASAP.

**Comment by slivne** [comment]
@raphaelsc we need the option to disable reshape on boot - the abort is ok if you want to fix something manually - but if you want todo automation its to complex - we need a command line option to allow disabling reshape.

**Comment by bhalevy** [comment]
> @raphaelsc we need the option to disable reshape on boot - the abort is ok if you want to fix something manually - but if you want todo automation its to complex - we need a command line option to allow disabling reshape.

@slivne we acted under @avikivity's guidelines here: https://github.com/scylladb/scylla/issues/7738#issuecomment-789642042
let's agree to agree on that before coding the option (which shouldn't be too complicated, but if we add it, we better also add a dtest for it)

**Comment by slivne** [comment]
@avikivity #7738 (comment) - the option to abort reshape is hard to use in automating upgrades.

During upgrade the cluster is running in mixed version. There are limitations during this period - and many users want to minimize it if possible.

So a run, attempting todo an upgrade will have to (skipping the backups etc)

Iterate on all nodes:
1. Upgrade node 4.y --> 4.x
2. Node starting 4.x - check logs if Reshape is executed
3. Run nodetool stop RESHAPE

Iterate on all nodes:
1. Restart node by node (allowing it to reshape)

With the flag we could do the following:

Iterate on all nodes:
1. Upgrade node 4.y --> 4.x (change config to bypass reshape)

Iterate on all nodes:
1. Update config remove bypass
2. Restart node


Its clear (at least to me) this will be simpler for users automating upgrades.

**Comment by raphaelsc** [comment]
what if we add a timeout for reshape? with a reasonable default like 5-10 minutes. with slow disks, like HDD, reshape will likely be interrupted due to timeout. with fast disks and not so many data to reshape, reshape will likely complete in time.  even if reshape is interrupted when timed out, the node will still benefit from all the completed reshape jobs, so progress isn't completely lost.

/CC @avikivity @slivne @bhalevy

**Comment by bhalevy** [comment]
> what if we add a timeout for reshape **on boot**? with a reasonable default like 5-10 minutes. with slow disks, like HDD, reshape will likely be interrupted due to timeout. with fast disks and not so much data to reshape, reshape will likely complete in time. even if reshape is interrupted when timed out, the node will still benefit from all the completed reshape jobs, so progress isn't completely lost.

Maybe moving reshape to the background is not such a bad option,
as long as it doesn't get in the way of regular compaction.

If we add a command-line option to control how reshape is done on boot, then it could be skipped entirely, or be moved to the background.  I don't think that doing that automatically after some timeout makes much sense.

Something like: `--reshape-compaction-mode=(enabled|disabled|background)`

**Comment by raphaelsc** [comment]
background reshape == regular compaction. so if we decide to bypass reshape on boot, that means: dear regular compaction, now it's your turn

**Comment by raphaelsc** [comment]
reshape on boot is only there because we want to compact data before node is online.

**Comment by bhalevy** [comment]
> background reshape == regular compaction. so if we decide to bypass reshape on boot, that means: dear regular compaction, now it's your turn

Not always.  For example, reboot after repair or compaction strategy change, or possibly upgrade (affecting the compaction strategy invariants)

**Comment by raphaelsc** [comment]
> > background reshape == regular compaction. so if we decide to bypass reshape on boot, that means: dear regular compaction, now it's your turn
> 
> Not always. For example, reboot after repair or compaction strategy change, or possibly upgrade (affecting the compaction strategy invariants)

Reboot after repair or compaction strategy change, means that strategy is far from its layout goal. 

Reshape == compact data until layout goal is achieved BEFORE node is online
Regular = compact data incrementally until layout goal is achieved WHILE node is online

That's why background reshape and regular are essentially the same thing.

**Comment by bhalevy** [comment]
ping @raphaelsc 

Also, anything else to do on this issue or can it be closed?
I believe we decided against a command line / configuration option and we're satisfied with the ability to stop reshaping via the api, right?

**Comment by tzach** [comment]
Also, https://github.com/scylladb/scylla/commit/0f50f8ec5f8e90e260bc4223f97ef01ee29f3884 does not fix this issue (not API) which is confusing.

**Comment by raphaelsc** [comment]
> ping @raphaelsc
> 
> Also, anything else to do on this issue or can it be closed? I believe we decided against a command line / configuration option and we're satisfied with the ability to stop reshaping via the api, right?

reshape stop API will be available in scylla 4.6, closing this issue...
